### PR TITLE
Introduce CodecID `00 01` to AvalancheJS

### PR DESCRIPTION
**Pull request by cgcardona**
## Introduction

As part of the Apricot hardfork we need to introduce codecID `00 01` to avalanchejs in a way which is backward compatible w/ codecID `00 00`. 

## Testing Scripts

We need to test building transactions manually, using components such as [BaseTx](https://github.com/ava-labs/avalanchejs/blob/master/src/apis/avm/basetx.ts#L27), as well as using the high level utility methods such as [buildBaseTx](https://github.com/ava-labs/avalanchejs/blob/master/src/apis/avm/api.ts#L721). Also we need to test codecID `00 00` as well as codecID `00 01` and confirm that they're interoperable.

### AVM

Each of these tests will work for both codecID `00 00` and `00 01`. Currently when you run them w/ no changes, other than adding the missing privkey, they default to codecID `00 00` which is the `LATESTCODEC` in the [AVM Constants](https://github.com/ava-labs/avalanchejs/blob/master/src/apis/avm/constants.ts#L7). In all [toBuffer](https://github.com/ava-labs/avalanchejs/blob/codecid-01/src/common/tx.ts#L95) calls we're also setting the `codecID` parameter's default value to `LATESTCODEC`. This enables us to flip the switch in one place, `AVMConstants.LATESTCODEC`, and update the latest codecID across all of AvalancheJS.

To test codecID `00 00` you simply leave the scripts as they are. `sign` now has an [optional 2nd parameter](https://github.com/ava-labs/avalanchejs/blob/codecid-01/src/apis/avm/tx.ts#L84) to set the codecID. Both `sign` and `issueTx` will each call `toBuffer` and correctly pass the codecID. 

```ts
// for codecID 00 00
const tx: Tx = unsignedTx.sign(xKeychain)
const id: string = await xchain.issueTx(tx)
console.log(id)
```

To test codecID `00 01`, first create a `codecID` variable w/ the value `1` and pass that to `sign` and `tx.toBuffer`. Next, convert the tx buffer to a cb58 encoded string. Lastly pass that to `issueTx`.

```ts
// for codecID 00 01
const codecID: number = 1
const tx: Tx = unsignedTx.sign(xKeychain, codecID)
const cb58EncodedTx: string = bintools.cb58Encode(tx.toBuffer(codecID))
const id: string = await xchain.issueTx(cb58EncodedTx)
console.log(id)
```

Each of the tests have the needed code for testing against both codecIDs at the bottom. Look for and comment/uncomment the following lines as needed.

```ts
// start uncomment for codecID 00 00
// stop uncomment for codecID 00 00
// start uncomment for codecID 00 01
// stop uncomment for codecID 00 01
```

#### Manual

Send AVAX in a `BaseTx`

* [avm-baseTx-avax](https://gist.github.com/cgcardona/06bb073dc0a8e865e5b773aa98652d42)

Send AVAX and ANT in a `BaseTx`

* [avm-baseTx-ant](https://gist.github.com/cgcardona/5a193aa415a67dcfe76fa43d702937ab)

Create an ANT

* [avm-createAssetTx-ant](https://gist.github.com/cgcardona/1ffa78db93ccd36825815751289a9484)

Create an NFT

* [avm-createAssetTx-nft](https://gist.github.com/cgcardona/b2a3cb899055444a0a512524c8afe372)

Mint an ANT

* [avm-operationTx-mint-ant](https://gist.github.com/cgcardona/b8e2bd0e038974ef19eea4a90f230ec3)

Mint an NFT

* [avm-operationTx-mint-nft](https://gist.github.com/cgcardona/7d9caf6bb0fb1a5d3a434890dc95ce8a)

Export AVAX from the X-Chain to the C-Chain or P-Chain

* [avm-exportTx-avax](https://gist.github.com/cgcardona/fad3be6c4829dc17f5216da6c82fdf51)

Export AVAX and ANT from the X-Chain to the C-Chain

* [avm-exportTx-ant](https://gist.github.com/cgcardona/3073e4c9c4e0aa0fad59fc7c0b3c1132)

Import AVAX from the C-Chain or the P-Chain to the X-Chain

* [avm-importTx-avax](https://gist.github.com/cgcardona/9916645b3ca25d47193b95a03dcc08d5)

Import AVAX and ANT from the C-Chain

* [avm-importTx-ant](https://gist.github.com/cgcardona/4061e9c4981696aa1b15acdfc4c04fef)

Still need to transfer NFTs.

#### High Level Utility Methods

Send AVAX in a `BaseTx`

* [avm-buildBaseTx-avax](https://gist.github.com/cgcardona/a7c7d3f8ba58eeafa9a7de908f288280)

Send an ANT in a `BaseTx`

* [avm-buildBaseTx-ant](https://gist.github.com/cgcardona/64ca8c7a2a70764c7ffe24247bfe171e)

Create an ANT

* [avm-buildCreateAssetTx-ant](https://gist.github.com/cgcardona/f927ba1705cc8794701f5d4c410deace)

Create an NFT

* [avm-buildCreateNFTAssetTx-nft](https://gist.github.com/cgcardona/29bf3869f046538b263b04f089c813bb)

Mint an ANT

* [avm-buildSECPMintTx-ant](https://gist.github.com/cgcardona/58fd90402cfb1983113cd0328afe5759)

Mint an NFT

* [avm-buildCreateNFTMintTx-nft](https://gist.github.com/cgcardona/f881d52f9d688aff0cdb1e8f862e8782)

Transfer an NFT

* [avm-buildNFTTransferTx-nft](https://gist.github.com/cgcardona/02bb60ecc0031722c9a5767a3631b6fb)

Import AVAX to the X-Chain from the C-Chain or the P-Chain

* [avm-buildImportTx-avax](https://gist.github.com/cgcardona/bae4fe38e7e688cc9933f5da3c52a9b3)

Export AVAX from the X-Chain to the C-Chain or the P-Chain

* [avm-buildExportTx-avax](https://gist.github.com/cgcardona/dff91dc66cf9ff2ebff6526294c7be96)

#### Manual

Import AVAX to the C-Chain from the X-Chain

* [evm-importTx-avax](https://gist.github.com/cgcardona/53a68e11957dd6fdbd0f77aaec0e5706)

Import AVAX and ANT to the C-Chain from the X-Chain

* [evm-importTx-ant](https://gist.github.com/cgcardona/bfecbee2b6cc69d3c1e568c6b9b19807)

Export AVAX to the X-Chain from the C-Chain

* [evm-exportTx-avax](https://gist.github.com/cgcardona/006028ecce7b22157179e2eb333d57c9)

Export AVAX and ANT from the C-Chain to the X-Chain

* [evm-exportTx-ant](https://gist.github.com/cgcardona/6924f0c8570e9a12cf2e7c361cab11d4)

#### High Level Utility Methods

Import AVAX to the C-Chain from the X-Chain

* [evm-buildImportTx-avax](https://gist.github.com/cgcardona/4427869819ef24436cba73d90c20214f)

Export AVAX to the X-Chain from the C-Chain

* [evm-buildExportTx-avax](https://gist.github.com/cgcardona/4427869819ef24436cba73d90c20214f)

#### Manual

Import AVAX to the P-Chain from the X-Chain

* [platformvm-importTx-avax](https://gist.github.com/cgcardona/ddb580e59bfa62832508c66416040576)

Export AVAX to the X-Chain from the P-Chain

* [platformvm-exportTx-avax](https://gist.github.com/cgcardona/73921a42c1bdfe118ac42ac11201440e)

Create a new subnet

* [platformvm-createSubnetTx](https://gist.github.com/cgcardona/34c5709a20374dceaeee6316fcc5175e)

Add a validator to the primary subnet

* [platformvm-addValidatorTx](https://gist.github.com/cgcardona/a7442871acf8466c09663eee6d4e32aa)

Add a delegator to the primary subnet

* [platformvm-addDelegatorTx](https://gist.github.com/cgcardona/240f3a8e1011e804d31510be91cccaa7)

#### High Level Utility Methods

Import AVAX to the P-Chain from the X-Chain

* [platformvm-buildImportTx-avax](https://gist.github.com/cgcardona/44b0856b0f1c0d1c8de861b31fadafe8)

Export AVAX to the X-Chain from the C-Chain

* [platformvm-buildExportTx-avax](https://gist.github.com/cgcardona/181d7985f048256fd23be9c7b157f1e4)

Add validator to the primary subnet

* [platformvm-buildAddValidatorTx-avax](https://gist.github.com/cgcardona/19f1a0681884faed08faa30bcd153d7a)

Add delegator to the primary subnet

* [platformvm-buildAddDelegatorTx-avax](https://gist.github.com/cgcardona/8454a5da55099f1cfbdcd17ce6f8dcca)

Create a new subnet

* [platformvm-buildCreateSubnetTx-avax](https://gist.github.com/cgcardona/cfc3461309ee9d8c3cbe098d70285acc)

 ## Avash

This can all be tested on localhost using [Avash](https://github.com/ava-labs/avash). You need to run avash on the `add-snow-epoch-first-transition-flag` branch.

```zsh
cd /path/to/avash
git checkout add-snow-epoch-first-transition-flag
go build
```

After re-building Avash next open the `five_node_staking.lua` script and add the `--snow-epoch-first-transition` flag which accepts as it's argument a timestamp that triggers the Apricot hardfork. If you put a timestamp in the future then you'll still be using codecID `00 00`. If you put a timestamp in the past then you'll be using codecID `00 01`.

```lua
-- five_node_staking.lua
cmds = {
    "startnode node1 --snow-epoch-first-transition=1710058317 --db-enabled=false --staking-enabled=true --http-port=9650 --staking-port=9651 --log-level=debug --bootstrap-ips= --staking-tls-cert-file=certs/keys1/staker.crt --staking-tls-key-file=certs/keys1/staker.key",
    "startnode node2 --snow-epoch-first-transition=1710058317 --db-enabled=false --staking-enabled=true --http-port=9652 --staking-port=9653 --log-level=debug --bootstrap-ips=127.0.0.1:9651 --bootstrap-ids=NodeID-7Xhw2mDxuDS44j42TCB6U5579esbSt3Lg --staking-tls-cert-file=certs/keys2/staker.crt --staking-tls-key-file=certs/keys2/staker.key",
    "startnode node3 --snow-epoch-first-transition=1710058317 --db-enabled=false --staking-enabled=true --http-port=9654 --staking-port=9655 --log-level=debug --bootstrap-ips=127.0.0.1:9651 --bootstrap-ids=NodeID-7Xhw2mDxuDS44j42TCB6U5579esbSt3Lg --staking-tls-cert-file=certs/keys3/staker.crt --staking-tls-key-file=certs/keys3/staker.key",
    "startnode node4 --snow-epoch-first-transition=1710058317 --db-enabled=false --staking-enabled=true --http-port=9656 --staking-port=9657 --log-level=debug --bootstrap-ips=127.0.0.1:9651 --bootstrap-ids=NodeID-7Xhw2mDxuDS44j42TCB6U5579esbSt3Lg --staking-tls-cert-file=certs/keys4/staker.crt --staking-tls-key-file=certs/keys4/staker.key",
    "startnode node5 --snow-epoch-first-transition=1710058317 --db-enabled=false --staking-enabled=true --http-port=9658 --staking-port=9659 --log-level=debug --bootstrap-ips=127.0.0.1:9651 --bootstrap-ids=NodeID-7Xhw2mDxuDS44j42TCB6U5579esbSt3Lg --staking-tls-cert-file=certs/keys5/staker.crt --staking-tls-key-file=certs/keys5/staker.key",
}

for key, cmd in ipairs(cmds) do
    avash_call(cmd)
end
```

After saving `five_node_staking.lua` then fire up avash and run the script.

```zsh
./avash
avash> runscript scripts/five_node_staking.lua
RunScript: Running scripts/five_node_staking.lua
RunScript: Successfully ran scripts/five_node_staking.lua
```

### Add missing Interfaces

Most of our API responses are generic `Promise<object>` which makes it challenging to debug what response to expect in your scripts. Ultimately we need to add interfaces for all of our API call parameters and responses but in this PR we're adding interfaces for any API calls which we're making during adding codecID `00 01` to AvalancheJS.

#### Common

* `iUTXOID`

#### AVM

* `iGetBalanceResponse`
* `iGetBalanceParams`
* `iAVMUTXOResponse`
* `iGetTxStatusParams`
* `iGetTxStatusResponse`
* `iCreateAddressParams`
* `iCreateFixedCapAssetParams`
* `iExportKeyParams`
* `iImportKeyParams`
* `iGetBalanceParams`
* `iGetBalanceResponse`
* `iGetTxStatusParams`
* `iGetTxStatusResponse`
* `iGetTxStatusResponse`
* `iMintParams`
* `iExportParams`
* `iExportAVAXParams`
* `iImportParams`
* `iImportAVAXParams`
* `iListAddressesParams`
* `iGetAllBalancesParams`
* `iBalance`
* `iGetAssetDescriptionParams`
* `iSendResponse`
* `iSendParams`

#### EVM

* `iEVMUTXOResponse`

#### PlatformVM

* `iGetBalanceResponse`
* `iGetBalanceParams`
* `iPlatformVMUTXOResponse`
* `iGetMinStakeResponse`

### AS-358

[Debug memo of `00 00 00 00`](https://ava-labs.atlassian.net/browse/AS-358)

### CODE REVIEW

File: src/apis/avm/api.ts [hunk 1]
```diff
@@ -874,8 +876,8 @@ export class AVMAPI extends JRPCAPI {
     throw new Error("Error - AVMAPI.buildImportTx: Invalid destinationChain type: " + (typeof sourceChain) );
   }
   
-  const avmUTXOResponse: AVMUTXOResponse = await this.getUTXOs(ownerAddresses, srcChain, 0, undefined)
-  const atomicUTXOs: UTXOSet = await avmUTXOResponse.utxos;
+  const avmUTXOResponse: iAVMUTXOResponse = await this.getUTXOs(ownerAddresses, srcChain, 0, undefined)
```
Comment (EvanJRichard) [inline src/apis/avm/api.ts:879]: I like the use of `i` to denote an interface, is that in our style guide?

File: src/apis/avm/credentials.ts [hunk 2]
```diff
@@ -25,12 +25,18 @@ export const SelectCredentialClass = (credid:number, ...args:Array<any>):Credent
 
 export class SECPCredential extends Credential {
   protected _typeName = "SECPCredential";
+  // TODO - Bump _typeID to *_CODECONE after Apricot
```
Comment (EvanJRichard) [inline src/apis/avm/credentials.ts:28]: I don't immediately see an easy way to do this but I may as well mention it, is there a way we could bump them all at once like how we're doing with `AVMConstants.LATESTCODEC`?

File: src/apis/avm/initialstates.ts [hunk 3]
```diff
@@ -85,28 +85,28 @@ export class InitialStates extends Serializable{
     return offset;
   }
 
-  toBuffer():Buffer {
+  toBuffer(codecID: number = AVMConstants.LATESTCODEC):Buffer {
```
Comment (EvanJRichard) [inline src/apis/avm/initialstates.ts:88]: nice styleguide compliance update here :+1:

File: src/apis/avm/inputs.ts [hunk 4]
```diff
@@ -22,7 +22,7 @@ const serializer = Serialization.getInstance();
  * @returns An instance of an [[Input]]-extended class.
  */
 export const SelectInputClass = (inputid:number, ...args:Array<any>):Input => {
-  if (inputid === AVMConstants.SECPINPUTID) {
+  if (inputid === AVMConstants.SECPXFERINPUTID || inputid === AVMConstants.SECPXFERINPUTID_CODECONE) {
```
Comment (EvanJRichard) [inline src/apis/avm/inputs.ts:25]: Similar to my comment on `*_CODECONE`, this is probably premature, but I wonder if there's an easy way to say "any sexpxferinputid"

File: src/apis/platformvm/importtx.ts [hunk 5]
```diff
@@ -88,7 +88,7 @@ export class ImportTx extends BaseTx {
    */
   toBuffer():Buffer {
     if(typeof this.sourceChain === "undefined") {
-      throw new Error("ImportTx.toBuffer -- this.sourceChain is undefined");
+      throw new Error(`ImportTx.toBuffer -- ${this.sourceChain} is undefined`);
```
Comment (EvanJRichard) [inline src/apis/platformvm/importtx.ts:91]: Would an error message of `ImportTx.toBuffer -- given sourceChain ${this.sourceChain} is undefined` be more helpful, or is that redundant?

File: src/common/credentials.ts [hunk 6]
```diff
@@ -120,7 +121,7 @@ export abstract class Credential extends Serializable{
 
   protected sigArray:Array<Signature> = [];
 
-  abstract getCredentialID():number;
+  abstract getCredentialID(codecID?:number):number;
```
Comment (EvanJRichard) [inline src/common/credentials.ts:124]: typescript question that I'm having a little trouble googling - does `codecID?:number` mean "maybe omitted, else a number"?

**Review by EvanJRichard — APPROVED** [review]
Looking good! :ship: Some nonblocking questions and comments, mostly for my own learning.

File: src/apis/avm/api.ts (hunk 1 above)
Comment (cgcardona) [reply src/apis/avm/api.ts:879]: Yes I added it to [our style guide](https://ava-labs.atlassian.net/wiki/spaces/HYD/pages/1243447307/AvalancheJS+Style+Guide#Interfaces) last week. Nice catch!

File: src/common/credentials.ts (hunk 6 above)
Comment (cgcardona) [reply src/common/credentials.ts:124]: Yes the `?` operator in arguments and object properties means it's optional. So in this case the `codecID` argument is optional but if it's implemented then it should be a `number`.

File: src/apis/platformvm/importtx.ts (hunk 5 above)
Comment (cgcardona) [reply src/apis/platformvm/importtx.ts]: Sure I think that makes sense. Updated.

File: src/apis/avm/outputs.ts [hunk 7]
```diff
@@ -99,15 +99,21 @@ export abstract class NFTOutput extends BaseNFTOutput {
  */
 export class SECPTransferOutput extends AmountOutput {
   protected _typeName = "SECPTransferOutput";
+  // TODO - Bump _typeID to *_CODECONE after Apricot
```
Comment (dasconnor) [inline src/apis/avm/outputs.ts:102]: todo

File: src/apis/avm/tx.ts [hunk 8]
```diff
@@ -50,6 +50,8 @@ export const SelectTxClass = (txtype:number, ...args:Array<any>):BaseTx => {
 export class UnsignedTx extends StandardUnsignedTx<KeyPair, KeyChain, BaseTx> {
   protected _typeName = "UnsignedTx";
   protected _typeID = undefined;
+  // TODO - Uncomment to set codecid
```
Comment (dasconnor) [inline src/apis/avm/tx.ts:53]: Looks like you need to set this for final merge

**Review by dasconnor — COMMENTED** [review]
Code looks good! Just resolve remaining todos or create Jira tickets for them (especially the post-apricot one). Let me know if you've already done that. I'll checkout the code tomorrow and test.

File: src/apis/avm/api.ts [hunk 9]
```diff
@@ -626,11 +634,11 @@ export class AVMAPI extends JRPCAPI {
...
    * @returns Returns a Promise<string> containing the status retrieved from the node
    */
-  getTxStatus = async (txid:string):Promise<string> => {
-    const params:any = {
+  getTxStatus = async (txid:string):Promise<iGetTxStatusResponse> => {
+    const params:iGetTxStatusParams = {
       txID: txid,
     };
-    return this.callMethod('avm.getTxStatus', params).then((response:RequestResponseData) => response.data.result.status);
+    return this.callMethod('avm.getTxStatus', params).then((response:RequestResponseData) => response.data.result);
```
[the remark this answers is not in the document]
Comment (cgcardona) [reply src/apis/avm/api.ts]: Agreed. Fixed

File: src/apis/avm/inputs.ts [hunk 10]
```diff
@@ -76,17 +76,24 @@ export abstract class AmountInput extends StandardAmountInput {
...
+  getInputID(codecID: number = AVMConstants.LATESTCODEC):number {
+    if(codecID === 0) {
+      return AVMConstants.SECPXFERINPUTID;
+    } else if (codecID === 1) {
+      this._typeID = AVMConstants.SECPXFERINPUTID_CODECONE;
+      return this._typeID;
+    }
   }
 
+  // TODO - Do we need to conditionally pass back SECPCREDENTIAL_CODECONE?
```
[the remark this answers is not in the document]
Comment (cgcardona) [reply src/apis/avm/inputs.ts]: Resolved.

File: src/apis/avm/ops.ts [hunk 11]
```diff
@@ -286,13 +287,19 @@ export class SECPMintOperation extends Operation {
...
+    } else if (codecID === 1) {
+      this._typeID = AVMConstants.SECPMINTOPID_CODECONE;
+      return this._typeID;
+    }
   }
 
   /**
    * Returns the credential ID.
    */
+  // TODO - Do we need to conditionally pass back SECPCREDENTIAL_CODECONE?
```
[the remark this answers is not in the document]
Comment (cgcardona) [reply src/apis/avm/ops.ts]: Resolved.

File: src/apis/avm/ops.ts [hunk 12]
```diff
@@ -538,13 +551,19 @@ export class NFTTransferOperation extends Operation {
...
+    } else if (codecID === 1) {
+      this._typeID = AVMConstants.NFTXFEROPID_CODECONE;
+      return this._typeID;
+    }
   }
 
   /**
    * Returns the credential ID.
    */
+  // TODO - Do we need to conditionally pass back NFTCREDENTIAL_CODECONE?
```
[the remark this answers is not in the document]
Comment (cgcardona) [reply src/apis/avm/ops.ts]: Resolved.

File: src/apis/avm/ops.ts [hunk 13]
```diff
@@ -389,22 +397,26 @@ export class NFTMintOperation extends Operation {
...
+    } else if (codecID === 1) {
+      this._typeID = AVMConstants.NFTMINTOPID_CODECONE;
+      return this._typeID;
+    }
   }
 
   /**
    * Returns the credential ID.
    */
+  // TODO - Do we need to conditionally pass back NFTCREDENTIAL_CODECONE?
```
[the remark this answers is not in the document]
Comment (cgcardona) [reply src/apis/avm/ops.ts]: Resolved.

File: src/apis/avm/ops.ts [hunk 14]
```diff
@@ -518,6 +530,7 @@ export class NFTMintOperation extends Operation {
  */
 export class NFTTransferOperation extends Operation {
   protected _typeName = "NFTTransferOperation";
+  // TODO - Bump _typeID to *_CODECONE after Apricot
```
[the remark this answers is not in the document]
Comment (cgcardona) [reply src/apis/avm/ops.ts:558]: Jira issue created https://ava-labs.atlassian.net/browse/AS-816

File: src/apis/avm/outputs.ts (hunk 7 above)
Comment (cgcardona) [reply src/apis/avm/outputs.ts:110]: Captured in JIRA ticket. Resolved.

File: src/apis/avm/credentials.ts (hunk 2 above)
Comment (cgcardona) [reply src/apis/avm/credentials.ts:28]: At the moment I don't see the obvious way either but I'm happy to brainstorm on it with you. I think this solution works for the apricot release.

File: src/apis/avm/inputs.ts (hunk 4 above)
Comment (cgcardona) [reply src/apis/avm/inputs.ts:25]: Similar answer. We can brainstorm on a better architecture after we get apricot out the door. For now I want to ship this and spend some more time extensively QAing to make sure there are no obvious regressions.

**Pull request merged**